cornice/validators/_colander.py

```diff
@@ -0,0 +1,63 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+from webob.multidict import MultiDict
+from colander import Invalid
+
+
+class SchemaError(Exception):
+    pass
+
+
+def validator(schema_class):
+    schema = schema_class()
+
+    def _validate(request):
+
+        if request.content_type == 'application/x-www-form-urlencoded':
+            body = request.POST.mixed()
+        elif not request.content_type == 'application/json':
+            body = request.body
+        else:
+            try:
+                body = request.json_body
+            except ValueError as e:
+                request.errors.add('body', '', 'Invalid JSON')
+                return
+            else:
+                if not hasattr(body, 'items'):
+                    request.errors.add('body', '', 'Should be a JSON object')
+                    return
+
+        cstruct = {'method': request.method,
+                   'url': request.url,
+                   'path': request.path,
+                   'body': body}
```

Comment:
Instead of inline code here, we could add a method `request.extract_as_dict()` (reusable with other validation libraries, easily testable, etc.)